spec/src/main/asciidoc/metrics-telemetry.asciidoc

```diff
@@ -0,0 +1,298 @@
+
+[cols="1,5"]
+|===
+| Name | `ft.invocations.total`
+
+| Type | `Counter`
+| Unit | None
+| Description | The number of times the method was called
+| Attributes
+a| * `method` - the fully qualified method name
+ * `result` = `[valueReturned\|exceptionThrown]` - whether the invocation returned a value or threw an exception
+ * `fallback` = `[applied\|notApplied\|notDefined]` - `applied` if fallback was used, `notApplied` if a fallback is configured but was not used, `notDefined` if a fallback is not configured
+|===
+
+=== Metrics added for `@Retry`
+
+[cols="1,5"]
+|===
+| Name | `ft.retry.calls.total`
+
+| Type | `Counter`
+| Unit | None
+| Description | The number of times the retry logic was run. This will always be once per method call.
+| Attributes
+a| * `method` - the fully qualified method name
+ * `retried` = `[true\|false]` - whether any retries occurred
+ * `retryResult` = `[valueReturned\|exceptionNotRetryable\|maxRetriesReached\|maxDurationReached]` - the reason that last attempt to call the method was not retried
+|===
+
+[cols="1,5"]
+|===
+| Name | `ft.retry.retries.total`
+
+| Type | `Counter`
+| Unit | None
+| Description | The number of times the method was retried
+| Attributes
+a| * `method` - the fully qualified method name
+|===
+
+=== Metrics added for `@Timeout`
+
+[cols="1,5"]
+|===
+| Name | `ft.timeout.calls.total`
+
+| Type | `Counter`
+| Unit | None
+| Description | The number of times the timeout logic was run. This will usually be once per method call, but may be zero times if the circuit breaker prevents execution or more than once if the method is retried.
+| Attributes
+a| * `method` - the fully qualified method name
+* `timedOut` = `[true\|false]` - whether the method call timed out
+|===
+
+[cols="1,5"]
+|===
+| Name | `ft.timeout.executionDuration`
+
+| Type | `Histogram`
+| Unit | Nanoseconds
+| Description | Histogram of execution times for the method
+| Attributes
+a| * `method` - the fully qualified method name
+|===
+
+=== Metrics added for `@CircuitBreaker`
+
+[cols="1,5"]
+|===
+| Name | `ft.circuitbreaker.calls.total`
+
+| Type | `Counter`
+| Unit | None
+| Description | The number of times the circuit breaker logic was run. This will usually be once per method call, but may be more than once if the method call is retried.
+| Attributes
+a| * `method` - the fully qualified method name
+ * `circuitBreakerResult` = `[success\|failure\|circuitBreakerOpen]` - the result of the method call, as considered by the circuit breaker according to the rules in <<circuitbreaker.asciidoc#circuit-breaker-success-failure,Configuring which exceptions are considered a failure>>
+ ** `success` - the method ran and was successful
+ ** `failure` - the method ran and failed
+ ** `circuitBreakerOpen` - the method did not run because the circuit breaker was in open or half-open state
+|===
+
+[cols="1,5"]
+|===
+| Name | `ft.circuitbreaker.state.total`
+
+| Type | `Gauge<Long>`
```

Comment:
In the part of the spec that refers to MP Metrics, the "Type" is a class from the MP Metrics API. Part of the TCK queries the MP Metrics API and looks for a metric of that type.

`Gauge<T>` is a type from MP Metrics, I'd expect this to be `DoubleGauge` (or maybe `DoubleCounter` since it's monotonically increasing).

However, I'm not sure whether it's appropriate to specify the type in the same way for Telemetry. I'm not super familiar with the OTel API, but it looks like the spec has several metric types, which can be either synchronous or asynchronous and the Java API breaks it down further depending on the type of the value (e.g. it has `DoubleCounter` and `LongCounter`).

I don't know whether we need to specify the exact type, or whether we can leave some of the descisions up to the implementation (e.g. if we know that a metric is a counter, do we need to specify if it's synchronous or asynchronous or can that safely be left up to the implementation).

Reply:
My thought is that anything that could affect the format of the output E.G. Double vs Long should be consistent across all implementations. Things like synchronous vs asynchronous can be left to implementers.